yarn-project/kv-store/src/lmdb-v2/write_transaction.ts

```diff
@@ -0,0 +1,314 @@
+    this.assertIsOpen();
+
+    const removeEntries = findInSortedArray(this.indexBatch.removeEntries, key, singleKeyCmp);
+    if (removeEntries && removeEntries[1] === null) {
+      return [];
+    }
+
+    const addEntries = findInSortedArray(this.indexBatch.addEntries, key, singleKeyCmp);
+    const results = await super.getIndex(key);
+
+    if (addEntries) {
+      merge(results, addEntries[1], Buffer.compare);
+      dedupeSortedArray(results, Buffer.compare);
+    }
+
+    if (removeEntries && Array.isArray(removeEntries[1])) {
+      removeAnyOf(results, removeEntries[1], Buffer.compare);
+    }
+
+    return results;
+  }
+
+  public override async *iterate(
+    startKey: Uint8Array,
+    endKey?: Uint8Array | undefined,
+    reverse?: boolean,
+    limit?: number,
+  ): AsyncIterable<[Uint8Array, Uint8Array]> {
+    yield* this.#iterate(
+      super.iterate(startKey, endKey, reverse),
+      this.dataBatch,
+      startKey,
+      endKey,
+      reverse,
+      limit,
+      (committed, toAdd) => (toAdd.length > 0 ? toAdd[0] : committed),
+      vals => vals[0],
+    );
+  }
+
+  public override async *iterateIndex(
+    startKey: Uint8Array,
+    endKey?: Uint8Array | undefined,
+    reverse?: boolean,
+    limit?: number,
+  ): AsyncIterable<[Uint8Array, Uint8Array[]]> {
+    yield* this.#iterate(
+      super.iterateIndex(startKey, endKey, reverse),
+      this.indexBatch,
+      startKey,
+      endKey,
+      reverse,
+      limit,
+      (committed, toAdd, toRemove) => {
+        if (toAdd.length > 0) {
+          merge(committed, toAdd, Buffer.compare);
+          dedupeSortedArray(committed, Buffer.compare);
+        }
+        if (toRemove.length > 0) {
+          removeAnyOf(committed, toRemove, Buffer.compare);
+        }
+        return committed;
+      },
+      vals => vals,
+    );
+  }
+
+  async *#iterate<T>(
+    iterator: AsyncIterable<[Uint8Array, T]>,
+    batch: Batch,
+    startKey: Uint8Array,
+    endKey: Uint8Array | undefined,
+    reverse: boolean = false,
+    limit: number | undefined,
+    merge: (committed: T, toAdd: Uint8Array[], toRemove: Uint8Array[]) => T,
+    map: (vals: Uint8Array[]) => T,
+  ): AsyncIterable<[Uint8Array, T]> {
+    this.assertIsOpen();
+
+    // make a copy of this in case we're running in reverse
+    const uncommittedEntries = [...batch.addEntries];
+    // used to check we're in the right order when comparing between a key and uncommittedEntries
+    let cmpDirection = -1;
+    if (reverse) {
+      cmpDirection = 1;
+      uncommittedEntries.reverse();
+    }
+
+    let uncommittedEntriesIdx = 0;
+    while (uncommittedEntriesIdx < uncommittedEntries.length) {
+      const entry = uncommittedEntries[uncommittedEntriesIdx];
+      // go to the first key in our cache that would be captured by the iterator
+      if (Buffer.compare(entry[0], startKey) !== cmpDirection) {
+        break;
+      }
+      uncommittedEntriesIdx++;
+    }
```

Comment:
Should be able to jump with the binary search here to avoid the linear search, right? Talking about unneeded optimizations...